File: Chapter13/Services/WorkoutService.cs
using Chapter13.Models;

namespace Chapter13.Services
{
    public class WorkoutService
    {
        private readonly List<Workout> _workouts = new List<Workout>();

        public IEnumerable<Workout> GetWorkouts() => _workouts;

        public Workout GetWorkoutById(int id) => _workouts.FirstOrDefault(w => w.Id == id);

        public void AddWorkout(Workout workout)
        {
            workout.Id = _workouts.Count + 1;
            _workouts.Add(workout);
        }

        public void UpdateWorkout(int id, Workout updatedWorkout)
        {
            var workout = _workouts.FirstOrDefault(w => w.Id == id);
            if (workout != null)
            {
                workout.Exercise = updatedWorkout.Exercise;
                workout.Duration = updatedWorkout.Duration;
                workout.Date = updatedWorkout.Date;
                workout.Notes = updatedWorkout.Notes;
            }
        }

        public void DeleteWorkout(int id)
        {
            var workout = _workouts.FirstOrDefault(w => w.Id == id);
            if (workout != null)
            {
                _workouts.Remove(workout);
            }
        }
    }

}

File: Chapter13/ClientApp/src/app/services/workout.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Workout {
  id: number;
  exercise: string;
  duration: number;
  date: Date;
  notes: string;
}

@Injectable({
  providedIn: 'root',
})
export class WorkoutService {
  private apiUrl = 'http://localhost:5124/api/workouts';

  constructor(private http: HttpClient) {}

  getWorkouts(): Observable<Workout[]> {
    return this.http.get<Workout[]>(this.apiUrl);
  }

  getWorkout(id: number): Observable<Workout> {
    return this.http.get<Workout>(`${this.apiUrl}/${id}`);
  }

  createWorkout(workout: Workout): Observable<Workout> {
    return this.http.post<Workout>(this.apiUrl, workout);
  }

  updateWorkout(id: number, workout: Workout): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}`, workout);
  }

  deleteWorkout(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
